lib/likelihood.c

```diff
@@ -0,0 +1,304 @@
+/*
+** Copyright (C) 2019 University of Oxford
+**
+** This file is part of msprime.
+**
+** msprime is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** msprime is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+ * Implementation of log-likelihood functions for full-ARG.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdlib.h>
+#include "likelihood.h"
+#include "msprime.h"
+
+int
+msp_increment_edge(tsk_treeseq_t *ts, tsk_edge_t *edge, tsk_node_t *parent, 
+                   tsk_node_t *child) {
+    int ret = 1;
+    ret = tsk_treeseq_get_edge(ts, edge->id + 1, edge);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->child, child);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->parent, parent);
+    if (ret != 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+int
+msp_unnormalised_log_likelihood_mut(tsk_treeseq_t *ts, double theta, 
+                                    double *lik) {
+    int ret = 1;
+    int it;
+    unsigned int j;
+    tsk_mutation_t mut;
+    tsk_size_t num_mutations = tsk_treeseq_get_num_mutations(ts);
+    tsk_id_t node_id;
+    double total_material = 0;
+    double branch_length;
+    tsk_tree_t *tree = NULL;
+    tsk_edge_t *edge = NULL;
+    tsk_node_t *parent = NULL;
+    tsk_node_t *child = NULL;
+    tree = malloc(sizeof *tree);
+    if (tree == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    edge = malloc(sizeof *edge);
+    if (edge == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    parent = malloc(sizeof *parent);
+    if (parent == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    child = malloc(sizeof *child);
+    if (child == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = tsk_treeseq_get_edge(ts, 0, edge);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->child, child);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_treeseq_get_node(ts, edge->parent, parent);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_tree_init(tree, ts, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    do {
+        total_material += (edge->right - edge->left) 
```

Comment:
I'd do this slightly differently:
```c
static double 
get_total_material(tsk_treeseq_t *ts)
{
    double ret = 0.0;
    tsk_edge_table_t *edges = ts->tables.edges;
    tsk_size_t j;

    for (j = 0; j < edges->num_rows; j++) {
         ret += (edges->right[j] - edges->left[j]) * (// etc)
    }
    return ret;
}
```

Saves quite a bit of faffing and is more efficient.